File: trunk/toolbox/src/main/scala/toolbox/SearchJar.scala
package toolbox
import toolbox.scalasupport._
import java.io.{File}
import java.util.regex.{Pattern, Matcher}
import java.util.zip.{ZipFile,ZipEntry}
object SearchJar {   
  var options = Map[String, String]()
  
  def main(argv: Array[String]): Unit = {
    var args = List[String]()
    for(s <- argv) s match {
      case "-h" | "--help" => printUsageAndExit
      case "-d" | "--debug" => options += ("debug"->"true")
      case "-f" | "--foldersearch" => options += ("foldersearch"->"true")
      case _ => args = s::args
    }
    
    if(args.size <=1) printUsageAndExit
    args = args.reverse
    
    val text = args(0)
    val jars = args.drop(1)
    val foldersearch = options.contains("foldersearch")
    
    debug("Searching: " + text)
    for(name <- jars){
      val dir = new File(name)
      if(dir.isFile && dir.getName.endsWith(".jar")){
        searchJar(text, dir)  
      }else{
        //search dir for jar, then match for text inside a jar.
        walkWithFile(dir){ subfile =>            
          if(subfile.getName.endsWith(".jar"))
            searchJar(text, subfile)  
        }
        
        if(foldersearch){
          //search dir for filename, then match text to filename(cases where classpath is a dir)
          walkWithDir(dir){ subdir =>
            val files = subdir.listFiles.filter{ _.getName.contains(text) }
            files.foreach{ f => println(getPathname(f)) }
          }
        }
      }
    }
  }

  def debug(s: String) = if(options.contains("debug")) println(s)

  def printUsageAndExit(){	
    println("Program to search text(entry name) inside a jar file.")
    println("Usage: scala SearchJar [-d|-f] text jarfile_or_dir [jarfile_or_dir ...]")
    println("   -d turn debug mode on.")
    println("   -f also search any folder for file name match.")
    exit(1)
  }
  
  def searchJar(text: String, jarfile: File): Unit = {
    debug("Searching: " + text + " in jarfile "+ getPathname(jarfile))
    val enu = new ZipFile(jarfile).entries
    while(enu.hasMoreElements){
      val entry = enu.nextElement.asInstanceOf[ZipEntry]
      val name = entry.getName
      if(name.contains(text))
        println(jarfile.getAbsolutePath + " " + name)
    }
  }
  
  /** walk a directory recursively and call func with each file found. */
  def walkWithFile(dir: File)(func: (File)=>Unit): Unit = {
    debug("Walk: " + dir)
    if(dir.isDirectory()) 
      dir.listFiles.foreach{ f => walkWithFile(f)(func) }
    else
      func(dir) //it's a file
  }
  /** walk a directory recursively and call func with each dir found. */
  def walkWithDir(dir: File)(func: (File)=>Unit): Unit = {
    debug("Walk: " + dir)
    if(dir.isDirectory()){
      func(dir) //it's a dir
      dir.listFiles.foreach{ f => walkWithDir(f)(func) }
    }
  }
  /** Get only the path upto where it first defined if not abosulte. */
  def getPathname(file: File): String = {
    if(file.isAbsolute) file.getAbsolutePath
    else{
      var f = file //init file path name
      val sb = new StringBuilder(f.getName)
      while(f.getParentFile != null){
        f = f.getParentFile
        sb.insert(0, f.getName+File.separator)
      }
      sb.toString
    }
  }
}

File: branches/javautilcmdline/trunk/src/main/jragonsoft/javautil/cmdtool/SearchJar.java
/*
 *	  5/4/2006 Zemian Deng
 * 
 *	  Licensed under the Apache License, Version 2.0 (the "License");
 *	  you may not use this file except in compliance with the License.
 *	  You may obtain a copy of the License at
 * 
 *		  http://www.apache.org/licenses/LICENSE-2.0
 * 
 *	  Unless required by applicable law or agreed to in writing, software
 *	  distributed under the License is distributed on an "AS IS" BASIS,
 *	  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *	  See the License for the specific language governing permissions and
 *	  limitations under the License.
 *																				 
 */

package jragonsoft.javautil.cmdtool;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Enumeration;
import java.util.Iterator;
import java.util.jar.JarFile;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.zip.ZipEntry;

import jragonsoft.javautil.support.GetOpt;
import jragonsoft.javautil.util.FileUtils;


/**
 * Utility program. Use -h or --help option in command line to see help page.
 * 
 * @author zemian
 * @version $Id: SearchJar.java 4 2006-03-16 15:27:19Z zemian $
 */
public class SearchJar {
	/** Description of the Field */
	public final static String JAR_EXT_PATTERN = "\\.jar$";

	private static boolean isDebug = false;

	private static boolean isRecursive = false;

	/** Description of the Method */
	public static void printExitHelp() {
		String help = "USAGE:"
				+ "\n  SearchJar [options] SearchText JarFile [JarDir ...]"
				+ "\n  Search filenames from one or more jar files. You may give dir as."
				+ "\n  as argument, and it will find all the jar files in it as input."
				+ "\n" + "\n  [options]" + "\n    -h      Help and version."
				+ "\n    -d      Show debug info."
				+ "\n    -r      Search recursively." + "\n" + "\nEXAMPLES:"
				+ "\n  $ # Assume using wrapper scripts and PATH is set"
				+ "\n  $ searchjar Servlet /opt/tomcat/common/lib" + "\n"
				+ "\nCREDITS:"
				+ "\n  ZMan Java Utility. <zemiandeng@gmail.com>"
				+ "\n  $Id: SearchJar.java 4 2006-03-16 15:27:19Z zemian $";

		System.out.println(help);
		System.exit(0);
	}

	/**
	 * The main program for the SearchJar class
	 * 
	 * @param args
	 *            The command line arguments
	 */
	public static void main(String[] args) {
		GetOpt opt = new GetOpt(args);

		//Process options
		if (opt.isOpt("h") || opt.getArgsCount() < 2) {
			printExitHelp();
		}
		isDebug = opt.isOpt("d");
		isRecursive = opt.isOpt("r");

		String searchStr = opt.getArg(0);
		ArrayList jarFiles = new ArrayList();

		//Get all jar files
		for (int i = 1; i < opt.getArgsCount(); i++) {
			File inputFile = new File(opt.getArg(i));
			if (inputFile.isDirectory()) {
				jarFiles.addAll(Arrays.asList(FileUtils.globFiles(inputFile,
						JAR_EXT_PATTERN, isRecursive)));
			} else if (inputFile.getName().toLowerCase().endsWith(".jar")) {
				Pattern p = Pattern.compile(JAR_EXT_PATTERN);
				Matcher m = p.matcher(inputFile.getName());
				if (m.find())
					jarFiles.add(inputFile);
			}
		}
		//debug("Seaching " + jarFiles);

		//Do searching on jar files
		int foundCount = 0;
		for (Iterator itr = jarFiles.iterator(); itr.hasNext();) {
			File item = (File) itr.next();
			debug("Seaching " + item.getAbsolutePath());
			try {
				Enumeration enu = new JarFile(item).entries();
				int checkCount = foundCount;
				while (enu.hasMoreElements()) {
					ZipEntry zipEntry = (ZipEntry) enu.nextElement();
					debug("zipEntry: " + zipEntry);
					if (zipEntry.getName().indexOf(searchStr) != -1) {
						System.out.println(item.getAbsolutePath() + " "
								+ zipEntry.getName());
						foundCount++;
						//break; //zdeng, Do not stop search even found first
						// occurance!
					}
				}
				debug("Found " + (foundCount - checkCount));
			} catch (IOException e) {
				//System.out.println(e); //ignore it for now.
				continue;
			}
		}

		if (foundCount == 0) {
			System.out.println("Can't find " + searchStr
					+ " in given jars or dirs.");
		}
	}

	public static void debug(String msg) {
		if (isDebug)
			System.out.println("[DEBUG] " + msg);
	}
}
